docs(s3): add an example to document how to upload a file to s3 (#368)
ForestAdmin/agent-nodejs: docs/agent-customization/examples/upload-file-to-s3.md

> This example show how to upload a file to an S3 bucket.

Reviewer: ```suggestion
This example shows how to upload a file to a S3 bucket.
```

To illustrate our example, imagine we have a User collection with an avatar string field. This field will store the s3

> url of the avatar.

Reviewer: ```suggestion
URL of the avatar.
```

Also, I'm not sure why this isn't on the same line

## Create an AWS service

You must install the [AWS SDK](https://docs.aws.amazon.com/sdk-for-javascript/v2/developer-guide/node-sdk-install.html).

```bash
npm install aws-sdk
```

> You can copy/past this custom AWS service in your agent to use it.

Reviewer: ```suggestion
You can copy/paste this custom AWS service in your agent to use it.
```

```javascript
class AwsS3Service {
  constructor({accessKeyId, secretAccessKey, bucket}) {
    this.accessKeyId = accessKeyId;
    this.secretAccessKey = secretAccessKey;
    this.bucket = bucket;
  }

  async upload(file, path) {
    const {Location} = await this.client
      .upload({
        Bucket: this.bucket,
        Key: `${path}/${file.name}`,
        Body: file.buffer,
        ContentType: file.mimeType,
        ACL: 'private',
      })
      .promise();

    return Location;
  }

  getUrlSignedUrlFromPath(path) {
    return this.client.getSignedUrl('getObject', {
      Bucket: this.bucket,
      Key: path,
      Expires: 600,
    });
  }

  get client() {
    return new AWS.S3({
      credentials: {
        accessKeyId: this.accessKeyId,
        secretAccessKey: this.secretAccessKey,
      },
    });
  }
}
```

Instantiate the AWS service.

```javascript
const config ={
  accessKeyId: '...',
  secretAccessKey: '...',
  bucket: '...',
};
const s3 = new AwsS3Service(config);
```

> ## Add a method to decode the date uri given by the client

Reviewer: ```suggestion
## Add a method to decode the data URI given by the client
```

> Our data uri is not conventional because we have added the file name.

Reviewer: ```suggestion
Our data uri are not conventional because we have added the file name.
```

```javascript
function parseDataUri(dataUri) {
  if (!dataUri) return null;

  const [header, data] = dataUri.substring(5).split(',');
  const [mimeType, ...mediaTypes] = header.split(';');
  const result = {mimeType, buffer: Buffer.from(data, 'base64')};

  for (const mediaType of mediaTypes) {
    const index = mediaType.indexOf('=');
    if (index !== -1)
      result[mediaType.substring(0, index)] = decodeURIComponent(mediaType.substring(index + 1));
  }

  return result;
}
```

## Customize the field

First, we need to create a new field which will sign the url of the avatar to send to the client.

```javascript
collection.addField('signedAvatar', {
    columnType: 'String',
    dependencies: ['avatar'],
    getValues: records =>
      records.map(record => {

>         // get the path stored in s3. The `pathname` is started with `/` that's why we need to remove it.

Reviewer: ```suggestion
        // get the path stored in s3. The `pathname` starts with a `/`, that's why we need to remove it.
```

        const path = decodeURI(new URL(record.avatar).pathname.substring(1));
        return s3.getUrlSignedUrlFromPath(path);
      })
    }
  )
```

Then, we replace the field writing to store the new s3 url.

```javascript
collection
  .replaceFieldWriting('signedAvatar', async (newAvatarFile, context) => {
    if (!newAvatarFile) return {avatar: null};

>     // compute an unique path to avoid colision when uploading the same file twice for a different record.

Reviewer: ```suggestion
    // compute an unique path to avoid collision when uploading the same file twice for a different record.
```
